staging/src/k8s.io/apiserver/pkg/cel/library/authz.go

```diff
@@ -256,12 +275,26 @@ var authzLibraryDecls = map[string][]cel.FunctionOpt{
 		cel.MemberOverload("decision_reason", []*cel.Type{DecisionType}, cel.StringType,
 			cel.UnaryBinding(decisionReason))},
 }
+var authzLibraryDeclsV1 = map[string][]cel.FunctionOpt{
+	"fieldSelector": {
+		cel.MemberOverload("authorizer_fieldselector", []*cel.Type{ResourceCheckType, cel.StringType}, ResourceCheckType,
+			cel.BinaryBinding(resourceCheckFieldSelector))},
+	"labelSelector": {
+		cel.MemberOverload("authorizer_labelselector", []*cel.Type{ResourceCheckType, cel.StringType}, ResourceCheckType,
+			cel.BinaryBinding(resourceCheckLabelSelector))},
+}
 
-func (*authz) CompileOptions() []cel.EnvOption {
-	options := make([]cel.EnvOption, 0, len(authzLibraryDecls))
-	for name, overloads := range authzLibraryDecls {
+func (a *authz) CompileOptions() []cel.EnvOption {
+	options := make([]cel.EnvOption, 0, len(authzLibraryDeclsV0))
+	for name, overloads := range authzLibraryDeclsV0 {
 		options = append(options, cel.Function(name, overloads...))
 	}
+	if a.version >= 1 {
+		// TODO(liggitt): only allow expressions with these to be persisted when AuthorizeWithSelectors feature is enabled?
```

Comment:
@jpbetz do you have a recommendation here? I'm not sure how feature gate enablement interacts with the versioned library registration

Reply:
I think this will be the first time feature gate a library.

I was imagining that we'd control it at the environment level (e.g. cel/environment/base.go) rather than at the library level.

The environment configuration doesn't support this case quite right though. So we'll have to modify it.

In particular, we need to make it possible to introduce a new library version (we can do that), and enable it for `StoredExpressions` (we can also do that), but don't make it enable automatically in the next release for `NewExpressions` (we can't do that).

In this use case, I think we also need to be able to `Extend` the base environment for the validation of the new API CEL fields so that they can immediately use the library version for `NewExpressions` _if_ the feature gate is enabled.  Conditionally extending the base environment using a feature gate is possible. But extending an environment to enable a different version of a library that's already in the base environment is not something we've had to do before. So we'll have to add support for that.

Does that rough direction sound right? I can help with the details tomorrow if we're aligned on that part.